src/wallet/wallet.cpp

```diff
@@ -4229,21 +4245,41 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(const std::string& walle
         success = DoMigration(*local_wallet, context, error, res);
     }
 
+    // In case of reloading failure, we need to remember the wallet dirs to remove
+    std::set<fs::path> wallet_dirs;
     if (success) {
-        // Migration successful, unload the wallet locally, then reload it.
+        // Migration successful, unload all wallets locally, then reload them.
         assert(local_wallet.use_count() == 1);
+        wallet_dirs.insert(fs::PathFromString(local_wallet->GetDatabase().Filename()).parent_path());
         local_wallet.reset();
         res.wallet = LoadWallet(context, wallet_name, /*load_on_start=*/std::nullopt, options, status, error, warnings);
         res.wallet_name = wallet_name;
-    } else {
+        success = res.wallet != nullptr;
+    }
+    if (success && res.watchonly_wallet) {
+        assert(res.watchonly_wallet.use_count() == 1);
+        std::string watchonly_wallet_name = res.watchonly_wallet->GetName();
+        wallet_dirs.insert(fs::PathFromString(res.watchonly_wallet->GetDatabase().Filename()).parent_path());
```

Comment:
In https://github.com/bitcoin/bitcoin/commit/6eb3df7c0bbbad0873f5600d3869ac33bad6ed5e:

Why this line?. The watch-only wallet datadir is added at line 4295.
Same for the solvable one.

Reply:
If the wallet fails to reload, we can still know to clean it up.